src/test/java/com/microsoft/azure/relay/HybridConnectionListenerTest.java

```diff
@@ -86,11 +181,39 @@ public void acceptHttpConnectionTest() throws IOException {
 		conn.setRequestProperty("ServiceBusAuthorization", tokenString);
 
 		assertEquals("Response did not have the expected response code.", status, conn.getResponseCode());
-		assertTrue("Listener failed to accept connections from sender in http mode.", checkHttpConnectionTask.join());	
+		assertTrue("Listener failed to accept connections from sender in http mode.", handlerExecuted.get());
+		
+		listener.close();
+	}
+	
+	@Test
+	public void listenerReconnectionTest() {
+		HybridConnectionListener listener = new HybridConnectionListener(CONNECTION_URI, tokenProvider);
+		AtomicBoolean handlerExecuted = new AtomicBoolean(false);
+		
+		listener.setConnectingHandler((ex) -> {
+			assertTrue("Exception should be ConnectionLostException.", ex != null && ex instanceof ConnectionLostException);
+			handlerExecuted.set(true);
+		});
+		
+		listener.openAsync(Duration.ofSeconds(15)).join();
+		ControlConnection controlConnection = listener.getControlConnection();
+		controlConnection.close(); // This bypasses listener.close(), simulating an unexptected close
```

Comment:
>close [](start = 20, length = 5)

Wait, I'm confused. What does calling controlConnection.close do? Will it reconnect after you call close?

Reply:
If it's possible for the controlConnection to be in a state where it doesn't think it should reconnect but the outer HCL thinks things are in a state where it will reconnect could that lead to or hide potential issues?

---
In reply to: [261787118](https://github.com/Azure/azure-relay-java/pull/28#discussion_r261787118) [](ancestors = 261787118)

Reply:
What do you mean? We rely heavily on the "closeCalled" flag from the outer Listener, so if we call close on the inner ControlConnection, we can bypass the closeCalled flag. As far as the outer listener is concerned, it's an unexpected close and will try to reconnect.

---
In reply to: [261788947](https://github.com/Azure/azure-relay-java/pull/28#discussion_r261788947) [](ancestors = 261788947,261787118)

Reply:
The outer listener never tries to reconnect.

---
In reply to: [261800522](https://github.com/Azure/azure-relay-java/pull/28#discussion_r261800522) [](ancestors = 261800522,261788947,261787118)

Reply:
Created a new "closeCalled" flag for ControlConnection as well so it wouldn't reconnect

---
In reply to: [262211154](https://github.com/Azure/azure-relay-java/pull/28#discussion_r262211154) [](ancestors = 262211154,261800522,261788947,261787118)